url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/misc/NumberOperations.java

```diff
@@ -0,0 +1,179 @@
+
+/**
+ * Arithmetic Operations as specified in chapter 1.7.
+ * 
+ * @author Christoph Beck
+ */
+public class NumberOperations {
+	private final static Long LONG_ZERO = Long.valueOf(0L);
+
+	private final static boolean isDotEe(String value) {
+		int length = value.length();
+		for (int i = 0; i < length; i++) {
+			switch (value.charAt(i)) {
+				case '.':
+				case 'E':
+				case 'e': return true;
+			}
+		}
+		return false;
+	}
+
+	private final static boolean isDotEe(Object value) {
+		return value instanceof String && isDotEe((String)value);
+	}
+
+	private static final boolean isFloatOrDouble(Object value) {
+		return value instanceof Float || value instanceof Double;
+	}
+
+	private static final boolean isFloatOrDoubleOrDotEe(Object value) {
+		return isFloatOrDouble(value) || isDotEe(value);
+	}
+
+	private static final boolean isBigDecimalOrBigInteger(Object value) {
+		return value instanceof BigDecimal || value instanceof BigInteger;
+	}
+
+	private static final boolean isBigDecimalOrFloatOrDoubleOrDotEe(Object value) {
+		return value instanceof BigDecimal || isFloatOrDoubleOrDotEe(value);
+	}
+
+	public static final Number add(TypeConverter converter, Object o1, Object o2) {
+		if (o1 == null && o2 == null) {
+			return LONG_ZERO;
+		}
+		if (o1 instanceof BigDecimal || o2 instanceof BigDecimal) {
+			return converter.convert(o1, BigDecimal.class).add(converter.convert(o2, BigDecimal.class));
+		}
+		if (isFloatOrDoubleOrDotEe(o1) || isFloatOrDoubleOrDotEe(o2)) {
+			if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+				return converter.convert(o1, BigDecimal.class).add(converter.convert(o2, BigDecimal.class));
+			}
+			return converter.convert(o1, Double.class) + converter.convert(o2, Double.class);
+		}
+		if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+			return converter.convert(o1, BigInteger.class).add(converter.convert(o2, BigInteger.class));
+		}
+		return converter.convert(o1, Long.class) + converter.convert(o2, Long.class);
+	}
+
+	public static final Number sub(TypeConverter converter, Object o1, Object o2) {
+		if (o1 == null && o2 == null) {
+			return LONG_ZERO;
+		}
+		if (o1 instanceof BigDecimal || o2 instanceof BigDecimal) {
+			return converter.convert(o1, BigDecimal.class).subtract(converter.convert(o2, BigDecimal.class));
+		}
+		if (isFloatOrDoubleOrDotEe(o1) || isFloatOrDoubleOrDotEe(o2)) {
+			if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+				return converter.convert(o1, BigDecimal.class).subtract(converter.convert(o2, BigDecimal.class));
+			}
+			return converter.convert(o1, Double.class) - converter.convert(o2, Double.class);
+		}
+		if (o1 instanceof BigInteger || o2 instanceof BigInteger) {
+			return converter.convert(o1, BigInteger.class).subtract(converter.convert(o2, BigInteger.class));
+		}
+		return converter.convert(o1, Long.class) - converter.convert(o2, Long.class);
+	}
+
+	public static final Number mul(TypeConverter converter, Object o1, Object o2) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 108-113 into a separate method.